auth/enforcer.go

```
@@ -0,0 +1,94 @@
+type PrimaryEOptionsIn struct {
+	fx.In
+	Options []basculehttp.EOption `group:"primary_bascule_enforcer_options"`
+	Logger  log.Logger
+}
+
+func ProvidePrimaryBasculeEnforcer() fx.Option {
+	return fx.Provide(
+		fx.Annotated{
+			Name: "primary_bearer_validator_principal",
+			Target: func() bascule.Validator {
+				return bascule.CreateNonEmptyPrincipalCheck()
+			},
+		},
+
+		fx.Annotated{
+			Name: "primary_bearer_validator_type",
+			Target: func() bascule.Validator {
+				return bascule.CreateValidTypeCheck([]string{"jwt"})
+			},
+		},
+		PrimaryCapabilityValidatorAnnotated(),
+		BasculeCapabilityMetricProvider{ServerName: "primary"}.Annotated(),
+		fx.Annotated{
+			Group: "primary_bascule_enforcer_options",
+			Target: func(in PrimaryBearerValidatorsIn) basculehttp.EOption {
+				in.Logger.Log(level.Key(), level.DebugValue(), xlog.MessageKey(), "building bearer rules option", "server", "primary")
+				validators := filterNilValidators([]bascule.Validator{in.Principal, in.Type, in.Capability})
+				if len(validators) < 1 {
+					in.Logger.Log(level.Key(), level.WarnValue(), xlog.MessageKey(), "providing nil bearer rules option", "server", "primary")
+					return nil
+				}
+				return basculehttp.WithRules("Bearer", bascule.Validators(validators))
+			},
+		},
+		fx.Annotated{
+			Group: "primary_bascule_enforcer_options",
+			Target: func() basculehttp.EOption {
+				return basculehttp.WithRules("Basic", bascule.CreateAllowAllCheck())
+			},
+		},
+		fx.Annotated{
+			Group: "primary_bascule_enforcer_options",
+			Target: func(in PrimaryBasculeMetricListenerIn) basculehttp.EOption {
+				return basculehttp.WithEErrorResponseFunc(in.Listener.OnErrorResponse)
+			},
+		},
+
+		fx.Annotated{
+			Name: "primary_alice_enforcer",
+			Target: func(in PrimaryEOptionsIn) alice.Constructor {
+				in.Logger.Log(level.Key(), level.DebugValue(), xlog.MessageKey(), "building alice enforcer", "server", "primary")
+				if anyNil(in.Options) {
```

Similar to the constructor setup, I'm not sure if it's possible for any options to be nil, and I think if we want to build an enforcer but can't, we should probably return an error.

---

I'll see if I can test if nil options are added in the list. I'll also check the nil cases.